### ISSUE TITLE
Patterns proposal syntactic ambiguity: switch statement vs expression

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/language#2138]
The [patterns proposal](https://github.com/dart-lang/language/blob/master/working/0546-patterns/patterns-feature-specification.md) introduces a new "switch statement" expression syntax, e.g.:

```dart
double calculateArea(Shape shape) =>
  switch (shape) {
    case Square(length: l) => l * l;
    case Circle(radius: r) => math.pi * r * r;
  };
```

This is problematic because a statement is allowed to begin with an expression, hence when parsing code like this:
```dart
void f(int i) {
  switch (i) {
    case 0 =>
```
we need to look ahead all the way to the `=>` token in order to know whether we are looking at a switch statement or a switch expression.  Because of the way the parser is implemented (it rewrites the token stream and invokes listener methods as it consumes tokens), doing this sort of lookahead carries a high cost in terms of implementation effort and maintenance.  (And, worse yet, if we accept the suggestion in https://github.com/dart-lang/language/issues/2126 of using `:` instead of `=>` in switch expressions, then in some cases no amount of lookahead will be able to resolve the ambiguity).

To avoid this ambiguity, I think we should disallow an expression statement starting with `switch` (in much the same way that we disallow an expression statement starting with `{` today).

Note that this means the user can't do this:
```dart
void f(int i) {
  switch (i) {
    ...
  }.method();
}
```
The workaround would be to enclose the switch statement in parentheses:
```dart
void f(int i) {
  (switch (i) {
    ...
  }).method();
}
```

**Comment by munificent** [comment]
> To avoid this ambiguity, I think we should disallow an expression statement starting with `switch` (in much the same way that we disallow an expression statement starting with `{` today).

I agree 100%. When I get a chance, I'll update the proposal to specify this.

**Comment by munificent** [comment]
This is fixed in 0760390bf64bda0076fc5e2f19cb34d774ff654e.
